### ISSUE TITLE
PRTE server crashes when multiple applications run back-to-back

### ISSUE THREAD

**Issue by wrrobin** [pmix/prrte#131]
## Background information
With multiple applications running back-to-back, the PRTE server crashes at random point. On a few occasions, the server just got stuck and the application launched did not terminate. I am testing this with the Sandia OpenSHMEM unit tests by running "make check" after the server is launched. The issue occurs much less frequently when the PRTE server is launched and terminated for each application separately. Detailed configuration and outputs are mentioned below. 

##### What version of PMIx are you using? (e.g., v1.2.5, v2.0.3, v2.1.0, git branch name and hash, etc.)
libevent 2.0.22-stable
hwloc 2.0.2
PMIx v3.1 (commit 7680895b0c5dec9b42206ddee35c80fb1683f6ca)
prte (PMIx Reference RTE) 3.0.0rc1

### Please describe the system on which you are running
* Operating system/version: CentOS Linux release 7.3.1611
* Computer hardware: Intel(R) Xeon(R) CPU E5-2699 v3 @ 2.30GHz
* Network type: Intel(R) Omni-Path 100  

-----------------------------

## Details of the problem

Here are the steps that are leading to this crash, assuming Sandia OpenSHMEM is downloaded and configured in `sandia-shmem-basedir`

```
prte &
[1] 7328
DVM Ready

cd sandia-shmem-basedir
make check
```

Below is the last part of the output that is collected while PRTE was run with "-d" flag.
 
```
[node1:03550] [[18102,0],0] orted:comm:process_commands() Processing Command: ORTE_DAEMON_DVM_CLEANUP_JOB_CMD
[node1:03550] [[18102,0],0] orted:comm:process_commands() Processing Command: ORTE_DAEMON_ADD_LOCAL_PROCS
[node1:03550] [[18102,0],0] Releasing job data for [INVALID]
[node1:03550] [[18102,0],0] Releasing job data for [INVALID]
[node1:03550] sess_dir_finalize: proc session dir does not exist
[node1:03550] sess_dir_finalize: job session dir does not exist
[node1:03550] sess_dir_finalize: jobfam session dir not empty - leaving
[node1:03550] sess_dir_finalize: jobfam session dir not empty - leaving
[node1:03550] sess_dir_finalize: top session dir not empty - leaving
[node1:03550] sess_dir_finalize: proc session dir does not exist
[node1:03550] sess_dir_finalize: job session dir does not exist
[node1:03550] sess_dir_finalize: jobfam session dir not empty - leaving
[node1:03550] sess_dir_finalize: jobfam session dir not empty - leaving
[node1:03550] sess_dir_finalize: top session dir not empty - leaving
[node1:03550] [[18102,0],0] orted:comm:process_commands() Processing Command: ORTE_DAEMON_DVM_CLEANUP_JOB_CMD
[node1:03550] sess_dir_finalize: proc session dir does not exist
[node1:03550] sess_dir_finalize: job session dir does not exist
[node1:03550] sess_dir_finalize: jobfam session dir not empty - leaving
[node1:03550] sess_dir_finalize: jobfam session dir not empty - leaving
[node1:03550] sess_dir_finalize: top session dir not empty - leaving
[node1:03550] [[18102,0],0] orted:comm:process_commands() Processing Command: ORTE_DAEMON_DVM_CLEANUP_JOB_CMD
[node1:03550] [[18102,0],0] orted:comm:process_commands() Processing Command: ORTE_DAEMON_ADD_LOCAL_PROCS
[node1:03550] [[18102,0],0] Releasing job data for [INVALID]
[node1:03550] [[18102,0],0] Releasing job data for [INVALID]
[node1:03550] pmix_ptl_base: send_msg: write failed: Bad address (14) [sd = 25]
*** Error in `prte': munmap_chunk(): invalid pointer: 0x00007f24d0053a28 ***
======= Backtrace: =========
/usr/lib64/libc.so.6(+0x7ada4)[0x7f24ddbeada4]
/home/rahmanmd/prrte-install-trial/pmix-3.1/pmix-install/lib/libpmix.so.2(pmix_ptl_base_send_handler+0x3a5)[0x7f24df354a46]
/home/rahmanmd/prrte-install-trial/libevent-2.0.22-stable/libevent-install/lib/libevent-2.0.so.5(event_base_loop+0x812)[0x7f24dee2be82]
/home/rahmanmd/prrte-install-trial/pmix-3.1/pmix-install/lib/libpmix.so.2(+0x94931)[0x7f24df2fb931]
/usr/lib64/libpthread.so.0(+0x7dc5)[0x7f24ddf38dc5]
/usr/lib64/libc.so.6(clone+0x6d)[0x7f24ddc6773d]
```

**Comment by rhc54** [comment]
Hi @wrrobin - does this occur after some application fails? Or is it after some number of applications run, with no app failing or returning a non-zero exit code?

I ask because we know that app failures or returns with non-zero codes can cause PRRTE to enter into an unstable state. We are working on that, but need specific reproducers to guide the work.

**Comment by wrrobin** [comment]
@rhc54 No, this occurs at random points with no applications failing prior to that. After a number of applications run and passed, it either crashes or gets stuck and no longer proceeds.

**Comment by wrrobin** [comment]
This has happened at random places for me. I am running a total of 84 applications back-to-back and it has occurred after 15-20 applications most of the time. However, once it could go almost up to 60 applications as well. The easiest re-producer would be to run a single application 100 times back-to-back without terminating the server in between.

**Comment by rhc54** [comment]
Okay - thanks! I'll take a look at it.

**Comment by davidozog** [comment]
Out of curiosity, I wrote a script to run our "Hello, World" test with 4 processes 1000 times.  It failed exactly 0 times...  So this problem may only manifest with more sophisticated communications than just `PMIx_Commit`/`PMIx_Fence`.  For me, the server crashes ~50 tests into our "make check" (around the `c11_test_shmem_g` test), but it can vary.

**Comment by wrrobin** [comment]
Hm. interesting. I have run hello world 100 times with 2 PEs and it always fails around 20th time. However, I have also added a sleep of 5 seconds in between each run and then it does not fail.

**Comment by rhc54** [comment]
Yeah, I've had trouble replicating it on my setup, but I only ran on one machine. I do have a few VMs I can tie together - will try that next week (surgery on Mon will slow me down). There has been a history of problems with short-lived apps. I thought we had cleaned those up, but maybe not completely.

I'm assuming it is safe to say that you are running prte across multiple machines in these tests?

**Comment by davidozog** [comment]
Good luck with your surgery today @rhc54!

For when you're back in business, I can actually reproduce this on a single machine with a few processes.  But I would hope that surgery recovery has considerably higher priority in the near future than reproducing this bug ;)

**Comment by rhc54** [comment]
Oh yes - definitely does! 😄  Things went very well and I am recovering just fine. Supposed to be back on my feet again within the net two days. Will likely start thinking seriously about things early next week.

**Comment by rhc54** [comment]
BTW: pass along your reproducing method so I can let it rattle around in my old fossil brain.

**Comment by davidozog** [comment]
Very glad to hear things went well.  Good luck recouping!

My apologies for such a long comment.  I blame all software dependencies!  

Anyway, here's how I set up my environment on my Ubuntu 14.0.1 laptop.  The setup @wrrobin describes above is slightly different a CentOS Omni-Path system.  

Today I made an attempt to update `pmix`,  `prrte`, and `SOS` to (hopefully) makes things easier.  I initially tried staying on master branches wherever possible, but that led to some problems.  For example, SOS needs to update its configury to use PMIx master ([SOS issue #812](https://github.com/Sandia-OpenSHMEM/SOS/issues/812)).  So, now I'm on the release candidates.

Here's my PMIx configuration on v3.1.0rc1.  

`./configure --prefix=$HOME/install/pmix-3.1.0rc1 --with-libevent=$HOME/src/libevent-2.0.22-stable/install --with-hwloc=$HOME/install/hwloc-1.11.11 --with-devel-headers --disable-visibility --enable-debug CFLAGS=-O0 CC=gcc`

Those builds of `libevent-2.0.22-stable` and `hwloc-1.11.11` have no extra configure flags or anything special.

Here's my PRRTE configuration on v3.0.0rc1:
`./configure --prefix=$HOME/install/prrte-v3.0.0rc1 --with-libevent=$HOME/src/libevent-2.0.22-stable/install --with-pmix=$HOME/install/pmix-v3.1.0rc1 --with-hwloc=$HOME/install/hwloc-1.11.11 CC=gcc`

Here's my SOS configuration (v1.4.2):
`./configure --prefix=$HOME/install/SOS-pmix-debug --with-ofi=$HOME/install/libfabric --with-pmix=$HOME/install/pmix --enable-error-checking --enable-picky --enable-debug CFLAGS=-O0 CC=gcc`

That's with a standard libfabric build on the v1.6.0 tag (no extra configure flags).

In a separate shell, I make sure my `PATH` and `LD_LIBRARY_PATH` are all set correctly, then run `prte -d`.

Then in the SOS build directory (same path setup as above) I do `make`', then `make check TESTS= -j` (to build the tests quickly), then do:
`make check TEST_RUNNER=prun NPES=1`

Note that this is just a single process, though the same thing happens with multiple procs.

I see that 30-50 tests pass then the tests either freeze and the server crashes with something like:
```
[a5e628e89608:44030] sess_dir_finalize: jobfam session dir not empty - leaving
[a5e628e89608:44030] sess_dir_finalize: top session dir not empty - leaving
[a5e628e89608:44030] [[37809,0],0] Releasing job data for [INVALID]
[a5e628e89608:44030] [[37809,0],0] Releasing job data for [INVALID]
*** Error in `prte': munmap_chunk(): invalid pointer: 0x00007fcb900a72c8 ***
Aborted
```

I do seem to get a core dump file, but after loading it, the backtrace is completely fubar and seems useless.  I'll dig a bit more though.

Enjoy!  Please let me know if and how I can help.

**Comment by rhc54** [comment]
Well, we have made some progress. Here is what I got when using the HEAD of the SOS, OFI, PRRTE, and PMIx repos as of this evening (Dec 29):
```shell
PASS: hello
PASS: pi
PASS: micro_unit_shmem
PASS: circular_shift
PASS: accessible_ping
PASS: max_reduction
PASS: big_reduction
PASS: to_all
PASS: strided_put
PASS: barrier
PASS: bcast
PASS: put1
PASS: get1
PASS: swap1
PASS: ping
PASS: pingpong
PASS: shmalloc
PASS: shmem_calloc
PASS: shrealloc
PASS: shmemalign
PASS: get_g
PASS: iput32
PASS: iput64
PASS: iput128
PASS: iput_short
PASS: iput_double
PASS: iput_float
PASS: iput_long
PASS: iput_longdouble
PASS: iput_longlong
PASS: sping
PASS: pingpong-short
PASS: ipgm
PASS: iput-iget
PASS: swapm
PASS: cswap
PASS: waituntil
PASS: atomic_inc
PASS: set_lock
PASS: test_lock
PASS: test_lock_cswap
PASS: fcollect64
PASS: bigput
PASS: bigget
PASS: ns
PASS: bcast_flood
PASS: lfinc
PASS: shmem_info
PASS: query_thread
PASS: asym_alloc
PASS: set_fetch
PASS: alltoall
PASS: alltoalls
PASS: c11_test_shmem_g
PASS: c11_test_shmem_get
PASS: c11_test_shmem_p
PASS: c11_test_shmem_put
PASS: c11_test_shmem_atomic_fetch
PASS: c11_test_shmem_atomic_set
PASS: c11_test_shmem_atomic_add
PASS: c11_test_shmem_atomic_inc
PASS: c11_test_shmem_atomic_and
PASS: c11_test_shmem_atomic_or
PASS: c11_test_shmem_atomic_xor
PASS: c11_test_shmem_atomic_swap
PASS: c11_test_shmem_atomic_cswap
PASS: c11_test_shmem_wait_until
PASS: c11_test_shmem_test
PASS: get_nbi
PASS: put_nbi
PASS: rma_coverage
PASS: collect
PASS: repeated_barriers
PASS: repeated_syncs
PASS: broadcast_active_set
PASS: reduce_active_set
PASS: collect_active_set
PASS: atomic_bitwise
PASS: nop_collectives
PASS: self_collectives
PASS: zero_comm
PASS: sync-size
PASS: shmem_ctx_pipelined_reduce
PASS: many-ctx
PASS: shmem_test
PASS: global_exit
FAIL: cxx_test_shmem_complex
FAIL: cxx_test_shmem_g
FAIL: cxx_test_shmem_get
FAIL: cxx_test_shmem_p
FAIL: cxx_test_shmem_put
FAIL: cxx_test_shmem_atomic_fetch
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/odls_base_default_fns.c at line 1749
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/odls_base_default_fns.c at line 1749
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/odls_base_default_fns.c at line 1749
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/odls_base_default_fns.c at line 1749
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/odls_base_default_fns.c at line 1749
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/odls_base_default_fns.c at line 1749
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/odls_base_default_fns.c at line 1749
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/odls_base_default_fns.c at line 1749
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/odls_base_default_fns.c at line 1749
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/odls_base_default_fns.c at line 1749
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
[localhost.localdomain:35131] [[22964,0],0] ORTE_ERROR_LOG: Not found in file base/state_base_fns.c at line 674
FAIL: cxx_test_shmem_atomic_set
FAIL: cxx_test_shmem_atomic_add
FAIL: cxx_test_shmem_atomic_inc
FAIL: cxx_test_shmem_atomic_and
FAIL: cxx_test_shmem_atomic_or
FAIL: cxx_test_shmem_atomic_xor
FAIL: cxx_test_shmem_atomic_swap
FAIL: cxx_test_shmem_atomic_cswap
FAIL: cxx_test_shmem_wait_until
FAIL: cxx_test_shmem_test
FAIL: hello_f
FAIL: complex_reductions_f
FAIL: set_fetch_f
FAIL: shmem_info_f
FAIL: mt_a2a
FAIL: mt_contention
FAIL: mt_membar
FAIL: mt_lock_test
FAIL: threading
FAIL: web
FAIL: thread_wait
FAIL: query_thread_funneled
FAIL: shmem_ctx
============================================================================
Testsuite summary for Sandia OpenSHMEM 1.4.2
============================================================================
# TOTAL: 115
# PASS:  86
# SKIP:  0
# XFAIL: 0
# FAIL:  29
# XPASS: 0
# ERROR: 0
============================================================================
See test/unit/test-suite.log
Please report to https://github.com/Sandia-OpenSHMEM/SOS/issues
============================================================================
```
It looks like we have a resource leak because when I subsequently try to run something:
```shell
$ prun -n 1 hostname
--------------------------------------------------------------------------
All nodes which are allocated for this job are already filled.
--------------------------------------------------------------------------
```
So PRRTE thinks that all available resources are occupied. This tells me that we aren't quite cleaning up completely, and it's likely the reason for the failures.

@ggouaillardet If you have a chance, could you try to track it down?
Ralph

**Comment by ggouaillardet** [comment]
@rhc54 I can do that after the holidays.

I noted a similar behavior but I thought pmix/pmix#1015 took care of that.
I was able to evidence the issue (not 100%, and I guess some race conditions are involved too) with the `bad_exit` program and 2 tasks (both `PMIx_Init()`, one `exit(1)` and the other `sleep(2); PMIx_Finalize()` and forcing `pmix_globals.notifications` with 8 rooms only. After a bunch of runs, all slots are filled though no task is running.

I realized I do not understand the rationale behind the caching event, and I might be missing some incorrect behaviors because of that. What kind of events have to be cached and why ? I made a patch to force the hotel capacity to zero rooms, and IIRC, everything hanged ... At first glance, it seems some events has to be cached before they are delivered (e.g `prun -n 1 hostname`) but some do not (e.g. `prun -n 2 bad_exit`). In the latter case, I am under the impression some events are cached but will never be delivered (e.g. their fate is to be evicted and released).

Could you please shed some light on this for me ?

**Comment by rhc54** [comment]
The key to using the event system lies in deciding which events need to go only to specific processes vs those that should be broadcast to anyone listening for them. Events targeting specific processes must be cached to avoid the race condition whereby the event arrives before the target has a chance to register for it. This is actually the typical case for ```prun``` as it doesn't know the nspace of its child job until after the spawn command completes. If the child job is fast (e.g., hostname or /bin/true), then the job completes and the "job terminated" event gets generated before ```prun``` has a chance to register for it.

My usual "rule of thumb" is to target events that deal with something specific to a job - e.g., job termination. I also target process termination events to their own namespace so their peer's can know about it. Only general environmental stuff is broadcast - e.g., a daemon going down. If apps need something more than that, then they need to generate it themselves - or we need to discuss why PRRTE should do it in general.

What I still need to do (I suspect - I don't recall if it was already done) is to remove the event from the cache once it has been delivered to all specified targets. This would reduce the number of cached events, and there is nothing to be gained by holding on to an event that can never be delivered again. The event system _used_ to do this, but I didn't check to ensure it still did after the revamp.

However, I doubt that's the source of this problem. This strikes me as more of a scenario where the job's resources aren't being properly returned to the host node. The error message is consistent with the case where the slot count gets out-of-sync and is usually caused by some object's ref count being off that results in the loops in state_dvm.c and state_orted.c that return resources not being properly executed.

In other words, I'm pretty sure this is a PRRTE problem and not a PMIx issue.

**Comment by ggouaillardet** [comment]
Thanks Ralph, I do appreciate the length explanation.

And yes, this is more likely a PRRTE issue than a PMIx one !

**Comment by rhc54** [comment]
FWIW: I played around a little bit with this and it looked to me like the process/slot accounting was working fine whenever the procs all exited successfully. However, things went out of whack as soon as a process abnormally terminated or exited with non-zero status. I no longer got any output from ```hostname``` and got error logs out of odls_base_support_fns and the state framework indicating that a job object couldn't be found.

So it looks like it may be something connected with the errmgr, though a quick glance thru the dvm component doesn't reveal anything to these old eyes.

**Comment by rhc54** [comment]
This now looks really good - I was able to run the complete SOS test suite to completion using the current head of PRRTE and PMIx

@wrrobin @davidozog Can you folks give it a try?

**Comment by wrrobin** [comment]
Thanks @rhc54. I will give it a try.

**Comment by wrrobin** [comment]
@rhc54 I am getting the following error when building from master. Is there anything I am missing?
```
In file included from /usr/include/string.h:633:0,
                 from /home/rahmanmd/prrte-install-trial/hwloc-2.0.2/hwloc-install/include/hwloc.h:59,
                 from ../opal/hwloc/hwloc-internal.h:28,
                 from orted/pmix/pmix_server_dyn.c:41:
orted/pmix/pmix_server_dyn.c: In function 'interim':
orted/pmix/pmix_server_dyn.c:564:41: error: 'PMIX_SPAWN_TOOL' undeclared (first use in this function)
         } else if (PMIX_CHECK_KEY(info, PMIX_SPAWN_TOOL)) {
                                         ^
orted/pmix/pmix_server_dyn.c:564:20: note: in expansion of macro 'PMIX_CHECK_KEY'
         } else if (PMIX_CHECK_KEY(info, PMIX_SPAWN_TOOL)) {
                    ^
orted/pmix/pmix_server_dyn.c:564:41: note: each undeclared identifier is reported only once for each function it appears in
         } else if (PMIX_CHECK_KEY(info, PMIX_SPAWN_TOOL)) {
                                         ^
orted/pmix/pmix_server_dyn.c:564:20: note: in expansion of macro 'PMIX_CHECK_KEY'
         } else if (PMIX_CHECK_KEY(info, PMIX_SPAWN_TOOL)) {
                    ^
make[2]: *** [orted/pmix/pmix_server_dyn.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/home/rahmanmd/prrte-install-trial/prrte/orte'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/rahmanmd/prrte-install-trial/prrte/orte'
make: *** [all-recursive] Error 1
```

**Comment by rhc54** [comment]
Are you sure you are building against PMIx master? It is defined there - remember you have to build/install PMIx master first and then point PRRTE to the install location.

**Comment by wrrobin** [comment]
ok.. I was using pmix-3.1. Will try the master then.

**Comment by wrrobin** [comment]
@rhc54 With the master, I could run all the SOS unit tests with no error. The issue seems to be resolved now. It took a while to finish running all the tests though. Seems to be a little slower compared to hydra.

**Comment by rhc54** [comment]
Good to hear about the issue. Couple of things might lead to that behavior vs hydra:
- A lot depends upon how the PMIx integration was done. IIRC, this initial cut was done by directly replacing PMI calls with their PMIx equivalent - yes? If so, we made minimal attempt to optimize that code path as the whole point of PMIx is to eliminate many of those calls. In contrast, hydra relies on them and spent much more effort optimizing them.
- Our focus on PRRTE has been towards creating a fully PMIx-enabled environment for development, not production. I believe the OMPI folks intend to use it for production, but they'll probably spend some time optimizing it.

I'll go ahead and close this for now - feel free to open a new one if something arises.
Ralph

**Comment by davidozog** [comment]
FWIW:  I also see all tests passing now (thanks!), and also notice that it seems slower than before.

Yes, I believe Ralph's 1st bullet is correct.  I could be crazy, but I think the unit tests that did pass before were a bit snappier before this patch, even with 1 process.  It's not too bad though...  I mention it just in case the performance becomes an issue in our test suite (which sadly, it already is... 😕).  Let me know if you want us to take some measurements, file an issue, or anything like that.

**Comment by wrrobin** [comment]
Here is a comparison for future reference:
Old version with the bug:
```
[rahmanmd@node1 SOS]$ time oshrun -np 2 test/unit/hello
Hello World from 1 of 2
Hello World from 0 of 2

real    0m0.511s
user    0m0.028s
sys     0m0.015s
```
New version with the fix:
```
[rahmanmd@node1 SOS]$ time oshrun -np 2 test/unit/hello
Hello World from 0 of 2
Hello World from 1 of 2

real    0m8.296s
user    0m0.024s
sys     0m0.014s
```

**Comment by rhc54** [comment]
Err...what is ```oshrun``` doing in there? Or did you just rename ```prun```?

**Comment by wrrobin** [comment]
Yes.. it is just a wrapper for `prun`. For example, configured SOS with `--with-oshrun-launcher=/home/rahmanmd/prrte-install-trial/prrte-3.0.0rc1/prrte-install/bin/prun` for the old version.

**Comment by rhc54** [comment]
Then something is clearly wrong. I just ran the exact same cmd on my woefully slow CentOS VM running on top of Mac OSX:
```shell
$ time prun -np 2 test/unit/hello
Hello World from 0 of 2
Hello World from 1 of 2

real	0m0.237s
user	0m0.005s
sys	0m0.006s
```
which is quite a bit faster than even the pre-patch numbers you got. I wonder if there is something out of sorts in your setup...do you have any stale rendezvous files (files starting with pmix) in your tmp directories? Any old prte daemons hanging around?

**Comment by wrrobin** [comment]
Oh. wow.. that's a huge difference. I do not see any such stale files in my tmp directories. Besides, before launching, I always do a `prun --terminate` to make sure that the previous instance is killed. Not sure, what could cause this difference. I will try a different machine to see whether I can reproduce your number. Just to verify, I am using the following versions:
libevent 2.0.22-stable
hwloc 2.0.2
PMIx master
prte (PMIx Reference RTE) master

**Comment by rhc54** [comment]
Yep, that matches my setup. Sadly, ```prun --terminate``` won't kill any lingering older ```prte``` copies hanging around - it only kills the last one started. Still, if you don't see any loose files hanging around the tmp directory (starting with "ompi" or "pmix"), then confusion over the master shouldn't be the issue.

Let me know what you find - we really ought to try and understand the source of this discrepancy.

**Comment by wrrobin** [comment]
Ok. I checked and could not find any file matching those prefixes. One more thing is that, with the new version, even `prun --terminate` takes about 7-8 seconds, whether or not it can terminate the previous instance. I will definitely check in another environment soon.

**Comment by rhc54** [comment]
Sounds like the TCP connection is taking a long time to complete. What OS are you using?

**Comment by wrrobin** [comment]
This is the same system I mentioned at the top.
Operating system/version: CentOS Linux release 7.3.1611
Computer hardware: Intel(R) Xeon(R) CPU E5-2699 v3 @ 2.30GHz
Network type: Intel(R) Omni-Path 100

 @davidozog Are you also seeing the slowdown on Neptune?

**Comment by rhc54** [comment]
It strikes me that there is one difference that might be involved here. You have Omni-Path on our machine while I have just TCP. There is some alpha-level code in PRRTE/PMIx for fabric support that would be getting called on a per-job basis. I'm wondering if perhaps we are seeing some impact of it here.

Try adding ```--pmca pnet ^opa``` to your ```prte``` and ```prun``` commands, or just set ```PMIX_MCA_pnet=^opa``` in the environment prior to issuing those commands. Let's see if that makes a difference.

**Comment by davidozog** [comment]
Thanks @wrrobin.  That's almost exactly what I'm seeing on my laptop as well, with both 1 and 2 PEs.

**Comment by rhc54** [comment]
@davidozog Very odd - what OS are you running on your laptop? Obviously, you wouldn't have OPA on it, so that might be a red herring.

@ggouaillardet Can you run this quick timing test described above and see what you get on your system?

**Comment by davidozog** [comment]
I'm seeing 8.5 seconds for hello-world on a docker image of Ubuntu 16.04 on a Mac OSX laptop.  I was just about to check the time pre-patch.  Any suggestions on how far back to go from the PRTE and PMIx heads?

**Comment by ggouaillardet** [comment]
I can observe similar delays on both CentOS7 and archlinux, and they roughly always equal `ptl_tcp_connection_wait_time * ptl_tcp_max_retries` (default is `4*2=8`, the delay change if I change these default values via the environment)

**Comment by rhc54** [comment]
I'm not entirely sure - @wrrobin what hashes did you use for the comparison?

**Comment by wrrobin** [comment]
I used these
libevent 2.0.22-stable
hwloc 2.0.2
PMIx v3.1 (commit 7680895b0c5dec9b42206ddee35c80fb1683f6ca)
prte (PMIx Reference RTE) 3.0.0rc1

**Comment by rhc54** [comment]
Ah, so you are seeing issues on the connection! Interesting - I wonder why I am not seeing them on my CentOS7 VM? Might be worth investigating the source of the connection delay.

**Comment by rhc54** [comment]
One other thing to check: are you failing over to the usock/ptl component?? It would indicate that the TCP stack is inoperable. Easiest way to check might be to set ``--pmca ptl ^tcp``` on your prun cmd line and see if the delay goes away.

**Comment by ggouaillardet** [comment]
`ptl/tcp` causes a delay because there is no `/tmp/pmix.sys.<hostname>` file.

```
$ PMIX_MCA_ptl_base_verbose=2 ~/local/prrte/bin/prun -n 1 hostname
[c7:05250] ptl:tcp: connecting to server
[c7:05250] ptl:tcp:tool looking for system server at /tmp/pmix.sys.c7
[c7:05250] WAITING FOR CONNECTION FILE
[c7:05250] WAITING FOR CONNECTION FILE
[c7:05250] ptl:tcp:tool searching for session server pmix.c7.tool
...
```



`prun --pmca ptl ^tcp` simply fails 

```
$ PMIX_MCA_ptl_base_verbose=2 ~/local/prrte/bin/prun -n 1 --pmca ptl ^tcp hostname
[c7:05565] [../../../../../../src/pmix/src/mca/ptl/usock/ptl_usock.c:108] connect to server
[c7:05565] [../../../../../src/prrte/orte/tools/prun/prun.c:777] PMIx Error: UNREACHABLE
```

interestingly, a dumb `touch /tmp/pmix.sys.<hostname>` had the delay gone away !
So I guess this issue has been unnoticed for a while because once upon a time, this file was created by pmix/prte

**Comment by ggouaillardet** [comment]
well, the delay has gone but I get this error message
```
[arch:00516] PMIX ERROR: PMIX_ERR_FILE_READ_FAILURE in file ../../../../../../src/pmix/src/mca/ptl/tcp/ptl_tcp.c at line 809
```

**Comment by rhc54** [comment]
Very interesting - let me poke at it a bit. I'm curious to see why my VM isn't seeing this problem.

**Comment by davidozog** [comment]
Wow, yeah -  I can confirm that doing `touch /tmp/pmix.sys.<hostname>` gives me about a 20x speedup.  Cool!

I also see this error:
```
[a5e628e89608:51431] PMIX ERROR: PMIX_ERR_FILE_READ_FAILURE in file ../../../../../src/mca/ptl/tcp/ptl_tcp.c at line 809
```

**Comment by rhc54** [comment]
The error message is due to the file not containing the expected information. I have two questions I need to investigate:

- why am I not hitting that problem on my VM?
- why isn't the file being created, or else why are we still looking for it?

One question for you guys, and I'll bet this is the underlying reason: are you starting prte with the ```--system-server``` flag? If not, try that and I'll bet the problem goes away.

**Comment by davidozog** [comment]
Yup, adding `--system-server` fixes the delay too.

**Comment by ggouaillardet** [comment]
@rhc54 you are on spot ! I did forget the `--system-server` flag when testing from my laptop (and I guess I forgot I always use it on my workstation ...). adding that flag was enough to had the problem gone away.

**Comment by rhc54** [comment]
Okay - thanks! Let me ponder the best path forward. Reason behind that behavior is that some people put the start of prte in a script followed immediately by their first prun - which then fails because it cannot find the rendezvous file. So we added some delay to give prte a chance to create the file.

However, it does mean we delay for each variant of the file (system level, proc level, etc.). Might be best if we first scan to see what files exist - if we find some, then go down the connection attempt path. If not, then delay and scan again.

Let me see what we can do that won't mess up the other folks.

**Comment by wrrobin** [comment]
Sounds good to me.

**Comment by davidozog** [comment]
Sounds good to me too.  Although, it's really no problem on our end to start `prte` with `--system-server` if you want to avoid making that change.  So as @jdinan would say: dealer's choice.

**Comment by rhc54** [comment]
With the changes, it launches just as fast either way - so choice is yours! 😄

**Comment by wrrobin** [comment]
Not on the same compute node. Multiple instances may run at the same time, but all these instances should be in different nodes.

**Comment by ggouaillardet** [comment]
Here is what I have been in mind for a while but did not get a chance to make a proof-of-concept : add a `--daemon` option to `prte`. If I oversimplify, the parent simply `fork()`, the child is the `prte` server and notify the parent when the DVM is ready. And then the parent `exit(0)` once the DVM is ready.

With this option, there would be no more need for a `prte &; sleep 2` or some retry mechanism in the `ptl/tcp` component.

any thoughts ?

**Comment by rhc54** [comment]
It's an interesting idea. I'm not sure how much effort it's worth putting into, though, as I'm not sure this is really that pressing a problem. Most people start ```prte``` separately and leave it running while they use it, sometimes for days before shutting it down. It's really only automated testers that seem to be experiencing this issue, and just documenting that you need to delay between ```prte``` and the first ```prun``` command might be adequate.

One thing I've had requested (but haven't done anything about yet) was to have prun look for the local daemon, and if it isn't found, then have prun fork it off and connect to it. It addresses the "I forget if I already have a prte running and don't want to get into the problem of having two of them around" issue. I suppose prun could always just return the "unreachable" error as it currently does, but a couple of people did ask for this "auto-bootstrap" behavior.

I only mention it because it somewhat mirrors what you describe. Again, not sure it's worth putting any priority on it. I don't have a strong opinion either way on either of these ideas.

**Comment by ggouaillardet** [comment]
got it, I will not give it a high priority, though I think this is gonna be needed for Open MPI integration. Unless we recomment/suggest to do it _a la_ `lamboot`, we should have a way to wrap both `prte` and `prun` into a new `mpirun` wrapper (that does not rely on `sleep()` to wait for the DVM).

**Comment by wrrobin** [comment]
Thanks @ggouaillardet. This will definitely help the automated testing scripts.

**Comment by rhc54** [comment]
I believe this has now been fixed - please let us know if not.
